File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    stack = []
    
    for i in number:
        while stack and k > 0 and stack[-1] < i:
            stack.pop()
            k -= 1
        stack.append(i)
    
    if k != 0:
        stack = stack[:-k]
        
    return ''.join(stack)

# 알고리즘
# number 값을 한 자리씩 순회
# stack에 값을 삽입
# 마지막 자리 값이 현재 값보다 작으면 비우고 대체
# k가 0이 될때까지

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        ArrayDeque<Character> stack = new ArrayDeque<>();
        
        for (int i = 0; i < number.length(); i++) {
            char c = number.charAt(i);
            
            while (!stack.isEmpty() && k > 0 && stack.peekLast() < c) {
                stack.pollLast();
                k--;
            }
            
            stack.offer(c);
        }
        
        // k가 남았다면 추가적으로 뒤에 부분 제거
        while (k > 0) {
            stack.pollLast();
            k--;
        }
        
        StringBuilder sb = new StringBuilder();
        for (char c: stack) {
            sb.append(c);
        }
        
        return sb.toString();
    }
}